protos/src/auth_service/convert.rs

```diff
@@ -0,0 +1,397 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use phnxtypes::{credentials, crypto, identifiers, messages, time};
+use tonic::Status;
+
+use crate::{
+    common::convert::{InvalidNonceLen, QualifiedUserNameError},
+    validation::{MissingFieldError, MissingFieldExt},
+};
+
+use super::v1::{
+    AsClientId, ClientCredential, ClientCredentialCsr, ClientCredentialPayload, ClientVerifyingKey,
+    ConnectionEncryptionKey, ConnectionPackage, ConnectionPackagePayload, CredentialFingerprint,
+    EncryptedUserProfile, ExpirationData, MlsInfraVersion, SignatureScheme,
+};
+
+impl From<identifiers::AsClientId> for AsClientId {
+    fn from(value: identifiers::AsClientId) -> Self {
+        let (user_name, client_id) = value.into_parts();
+        Self {
+            user_name: Some(user_name.into()),
+            client_id: Some(client_id.into()),
+        }
+    }
+}
+
+impl TryFrom<AsClientId> for identifiers::AsClientId {
+    type Error = AsClientIdError;
+
+    fn try_from(proto: AsClientId) -> Result<Self, Self::Error> {
+        Ok(Self::new(
+            proto
+                .user_name
+                .ok_or_missing_field("user_name")?
+                .try_into()?,
+            proto.client_id.ok_or_missing_field("client_id")?.into(),
+        ))
+    }
+}
```

Comment:
Would it be worth at some point writing a macro for these `try_from` implementations? We're bound to have a lot of them. The same goes for the `From` implementations.

Reply:
Good question. Slowly I definitely see some patterns emerging. Now we have generic Keys and Ciphertext types, we could provide those implementations either via a macro or via a trait.

How about doing a sweep over the conversions after all services are implemented and trying to factor out those patterns?

Reply:
Sounds like a plan!